### ISSUE TITLE
Installation fails with FreeTDS 1.0 on OSX

### ISSUE THREAD

**Issue by andrewmwhite** [pymssql/pymssql#432]
Error message:
```
clang -fno-strict-aliasing -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -I/usr/local/include -I/usr/local/var/pyenv/versions/2.7.11/include/python2.7 -c _mssql.c -o build/temp.macosx-10.11-x86_64-2.7/_mssql.o -DMSDBLIB
_mssql.c:18783:15: error: use of undeclared identifier 'DBVERSION_80'
    __pyx_r = DBVERSION_80;
              ^
1 error generated.
error: command 'clang' failed with exit status 1
```

[Full log of failed install command](https://github.com/pymssql/pymssql/files/287439/pymssql.txt)

Reverting to FreeTDS 0.91 via `brew unlink freetds; brew install homebrew/versions/freetds091` fixes the issue as a work-around.

**Comment by nunoh** [comment]
Yep, same here. Used a very similar fix, but instead with `uninstall` rather than `unlink`.

**Comment by gdelgado** [comment]
+1

Same issue, but the above solutions has not worked for me.

**Comment by jacobbridges** [comment]
Same issue on Ubuntu. Apparently the apt sources have linked free-tds 1.0 as the current stable version so pymssql stopped working.

**Comment by ramiro** [comment]
> Same issue on Ubuntu. Apparently the apt sources have linked free-tds 1.0 as the current stable version so pymssql stopped working.

This is strange. Ubuntu (and Debian) are stuck on FreeTDS 0.91:

* http://packages.ubuntu.com/search?keywords=freetds
* https://packages.debian.org/search?keywords=freetds

There was a third party PPA with FreeTDS packages for Ubuntu which carried 0.95. IIRC it's this one:

https://launchpad.net/~jamiewillis/+archive/ubuntu/freetds?field.series_filter=trusty

What APT repos are you using in order to get 1.0?

BTW we are asking for testing/feedback on binary wheels for Linux we are publishing for the first time, which include static version of FreeTDS 0.95.x and we plan to release in the next few days as version 2.1.3 of pymssql:

https://groups.google.com/d/msg/pymssql/rMEJAaJhX1c/rSwSuejeNwAJ

**Comment by freddy77** [comment]
Just use DBVERSION_71. See http://www.freetds.org/userguide/choosingtdsprotocol.htm#AEN910. Also described in the NEWS file at https://github.com/FreeTDS/freetds/blob/master/NEWS.
DBVERSION_ are preprocessor macros so you can check if \#ifdef DBVERSION_80.
Actually version 7.1 refers to the protocol used by mssql 2000.

**Comment by tsoukw** [comment]
yes ,it's good for me, thank you

**Comment by jh86** [comment]
Will we see this fix in a new PyPI release soon?

**Comment by bkanuka** [comment]
@jh86 until this is in PyPI, you can install from github: `pip install git+https://github.com/pymssql/pymssql.git`

**Comment by daniloakamine** [comment]
it works, but in El Capitan I had some problems.
when I tried to install `freetds091` I received the error:
`configure: error: cannot run C compiled programs.`.
 
I ran `brew doctor` and I found out that my Xcode 6.1.1 was outdated.
So in that case, you will need update your Xcode or there is a simple solution, run `xcode-select --install`.
After this, try install `freetds091` again.

**Comment by mcescalante** [comment]
Because 2.2.0 still hasn't made it to PyPI as of this comment, the following command from @bkanuka still works and installed smoothly without a single error:

`pip install git+https://github.com/pymssql/pymssql.git`

**Comment by wesseljt** [comment]
thanks!

**Comment by mideeks** [comment]
Any chance we can get a 2.2.0/2.1.4 release soon so we can build on macOS without having to pull from git?

**Comment by giaosudau** [comment]
In my case I need to do one more step `brew link --force freetds@0.91` and it works perfect.

**Comment by drorata** [comment]
I work with macports. I had to do the following:

1. Remove installed older version of `freetds`
2. Checkout the macports repo `git clone --single-branch https://github.com/macports/macports-ports.git`
3. Checkout an older version, where the port is with an older version of `freetds`. In particular: `git checkout fb32c4486dab8bc03409927f0d73a8ac3e37cada`
4. Lastly, I ran `sudo port install` from `~/macports-ports/databases/freetds`

**Comment by anilpai** [comment]
The complete process is here.

> 
http://gree2.github.io/python/setup/2017/04/19/python-instal-pymssql-on-mac

**Comment by jzaleski** [comment]
Until `v2.2.0` has been blessed/released, as it is still in development, where the core issue resolved, you can use the following tag on our fork here: https://github.com/wayfair/pymssql/releases/tag/v2.1.3.0.0.1 -- it _has_ the necessary patch to resolve the original issue.

**Comment by nickmachairas** [comment]
@jzaleski Installation froze on my Mac with the dowloaded tar of release 'v2.1.3.0.0.1: Preparing for a 2.1.3.0.0.1 release' and `pip install`. But it worked with `easy_install`. I didn't find this posted anywhere else, hopefully it will help others with the same problem.

**Comment by adriannye** [comment]
With 2.1.3 I get the DBVERSION error.
I was previously successful pip installing the master branch but it no longer works.
Pip installing release 2.1.3.0.0.1  fails for me with the downloaded tar on MacOS Sierra with the error:

File "/usr/local/opt/pyenv/versions/3.4.3/envs/api/lib/python3.4/site-packages/setuptools/command/build_ext.py", line 46, in <module>

        class build_ext(_build_ext):

    TypeError: str() argument 2 must be str, not tuple

**Comment by jzaleski** [comment]
@nickmachairas could you please give me a bit more context on the issue? I think that _may_ be unrelated. But I'd be happy to take a look and help you work through it!

**Comment by jzaleski** [comment]
@adriannye just to confirm, you're on Python `3.4.3`? Can you give me slightly more of the stack-trace? I'd be glad to help you work through this.

**Comment by adriannye** [comment]
Thanks jzaleski, yes I'm on python 3.4.3 (pyenv) on Sierra.  There's a good chance the problem is related to pyenv as I just started using that, but ever other package is building fine.   About a month ago I successfully installed the pymssql master using py3.6 installed manually.   I have freetds 0.91 and setuptools 12.0.5.

The full traceback is here:
https://pastebin.com/4yAVVgMV

**Comment by adriannye** [comment]
I gave up on pyenv and brew installed python3 and the latest master works with that.  So the problem was definitely related to pyenv.

**Comment by jzaleski** [comment]
Glad to hear it @adriannye -- Please let us know if you need help w/ anything else. Good luck!

**Comment by nickmachairas** [comment]
Thank you @jzaleski, it is clear then that my issue was caused by something else. Thanks again!

**Comment by drorata** [comment]
After upgrading to macOS High Sierra, I tried to redo the steps in my [older comment](https://github.com/pymssql/pymssql/issues/432#issuecomment-295702344) but it didn't work. The `sudo port install` returned:

```
Error: Unable to execute port: Could not open file: ~/macports-ports/databases/freetds/Portfile
```

Any ideas?

**Comment by jzaleski** [comment]
It has been quite some time since I've used MacPorts, but it would appear
that there is an issue specific to the `freetds` port you are trying to
install. Is there any chance you could try installing freetds by another
means (I have validated that the contents of this PR work from `0.9*` to
current)?

**Comment by drorata** [comment]
@jzaleski I don't think the problem is due to `macports`. Using [this commit](https://github.com/pymssql/pymssql/commit/9eca0b3bf2729f203e7088a6e7de4cc3dc4e7d52) I have the expected behavior on the mac but it breaks inside the container (based on [this image](https://github.com/docker-library/python/blob/cf179e4a7b442b29d85f521c2b172b89ef04beef/3.6/jessie/Dockerfile)).

I don't think I will be able to try out non-macports based solution. I am sorry.

**Comment by drorata** [comment]
I will try to summarize the situation I am facing.

I consider two platforms:

1. **macOS** where `freetds` version 1.00.67 is installed using macports.
2. Docker **container** where `freetds` is installed using 
    ```
    RUN apt-get update && apt-get install -y \
      freetds-dev \
      python-dev \
      build-essential \
      && apt-get clean -y \
    ```
    I don't know how to check what version is it... The image's base is `python:3`

On the other hand I consider two versions of `pymssql`
1. **dev** as available in [this commit](https://github.com/pymssql/pymssql/tree/9eca0b3bf2729f203e7088a6e7de4cc3dc4e7d52).
2. **2.1.3** as pointed by [this commit](https://github.com/pymssql/pymssql/commit/9de15f5773725e4acafb86e1b97eb2bfe58554c4)

The problem I witness is twofold:

2. Inside a container I can install both version, *but* I can connect to the server only with the installation of 2.1.3. When trying to connect using the *dev* installation I get the following error: 

    ```
    Traceback (most recent call last):
      File "src/pymssql.pyx", line 636, in pymssql.connect
      File "src/_mssql.pyx", line 1952, in _mssql.connect
      File "src/_mssql.pyx", line 673, in _mssql.MSSQLConnection.__init__
      File "src/_mssql.pyx", line 1680, in _mssql.maybe_raise_MSSQLDatabaseException
     _mssql.MSSQLDatabaseException: (20002, b'DB-Lib error message 20002, severity 9:\nAdaptive Server connection failed\n')
    ```
1. Installing on macOS: I cannot install `2.1.3`. When trying `pip install pymssql` I get:
 
   ```
    Collecting pymssql
      Using cached pymssql-2.1.3.tar.gz
    Building wheels for collected packages: pymssql
      Running setup.py bdist_wheel for pymssql ... error
      Complete output from command /Users/drorata/anaconda3/envs/test-pymssql/bin/python -u -c "import setuptools, tokenize;__file__='/private/var/folders/6z/h5jqtwss0m13gg40vv06gs_40000gn/T/pip-build-q457zka4/pymssql/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" bdist_wheel -d /var/folders/6z/h5jqtwss0m13gg40vv06gs_40000gn/T/tmp2g3qg71tpip-wheel- --python-tag cp36:
      setup.py: platform.system() => 'Darwin'
      setup.py: platform.architecture() => ('64bit', '')
      setup.py: platform.libc_ver() => ('', '')
      setup.py: Detected Darwin/Mac OS X.
          You can install FreeTDS with Homebrew or MacPorts, or by downloading
          and compiling it yourself.

          Homebrew (http://brew.sh/)
          --------------------------
          brew install freetds

          MacPorts (http://www.macports.org/)
          -----------------------------------
          sudo port install freetds

      setup.py: Not using bundled FreeTDS
      setup.py: include_dirs = ['/opt/local/include', '/opt/local/include/freetds']
      setup.py: library_dirs = ['/usr/local/lib', '/opt/local/lib']
      running bdist_wheel
      running build
      running build_ext
      building '_mssql' extension
      creating build
      creating build/temp.macosx-10.7-x86_64-3.6
      gcc -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -I/Users/drorata/anaconda3/envs/test-pymssql/include -arch x86_64 -I/Users/drorata/anaconda3/envs/test-pymssql/include -arch x86_64 -I/opt/local/include -I/opt/local/include/freetds -I/Users/drorata/anaconda3/envs/test-pymssql/include/python3.6m -c _mssql.c -o build/temp.macosx-10.7-x86_64-3.6/_mssql.o -DMSDBLIB
      _mssql.c:18814:15: error: use of undeclared identifier 'DBVERSION_80'
          __pyx_r = DBVERSION_80;
  ```

**Comment by thongly** [comment]
I needed this

`brew link --force freetds@0.91`

-credit to http://gree2.github.io/

**Comment by ManuelGysin** [comment]
This problem is also in fedora 27.

**Comment by jzaleski** [comment]
We are in the process of pulling together a `2.1.4` wheel currently it will include this fix and others.

As an aside to address an earlier comment, depending on your hardware, it can take some time (minutes) to build the C-extension. Before we dig in further, please confirm that it isn't just taking a long time to compile. Thanks!

**Comment by kparaju** [comment]
@jzaleski any updates on `2.1.4`?

**Comment by jpgimenez** [comment]
@kparaju check 2.1.4 milestone https://github.com/pymssql/pymssql/milestone/9

**Comment by jfahrenkrug** [comment]
The complete steps that worked for me were:

    brew unlink freetds
    brew install freetds@0.91
    brew link --force freetds@0.91
    brew install pymssql

Thanks for your help!

**Comment by HongJheLi** [comment]
I have to do ... 
brew link `--overwrite` --force freetds@0.91

**Comment by fanghaolei** [comment]
```
brew unlink freetds
brew install freetds@0.91
brew link --overwrite --force freetds@0.91
pip install pymssql
```

This worked for me. I have python 3.6

**Comment by erg** [comment]
Still having this problem on macOS, none of these solutions worked anymore except this:

``sudo pip3 install git+https://github.com/pymssql/pymssql.git``

**Comment by AlexHagerman** [comment]
@erg can you link me to your open issue or post on the mailing list? When you say you are still having this problem are you running into the DBVERSION issue that is what this ticket was originally opened for or the sqlfront.h not found issue on ticket 124 where you cross posted?

**Comment by erg** [comment]
@AlexHagerman I hit both issues that I posted on by following various suggested fixes like installing ``freetds@0.91``. The only one that fixed my pymssql yesterday was installing from git.

I have hit this issue repeatedly over the last two years and installing pymssql is always a pain on a new system. I'd be happy to help debug it but looks like you have it figured out with the latest git version. Perhaps you could make a release so when people install ``pymssql`` from a ``requirements.txt`` file on macOS they will have the fix.
